File: rec-accessor/src/main/java/net/kimleo/rec/accessor/RecordWrapper.java
package net.kimleo.rec.accessor;

import net.kimleo.rec.concept.Indexible;
import net.kimleo.rec.concept.Mapped;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class RecordWrapper<T> implements Indexible<T>, Mapped<T> {

    private final Map<String, Integer> fieldNames;

    private final Indexible<T> record;

    public RecordWrapper(Map<String, Integer> fieldNames, Indexible<T> record) {
        this.fieldNames = fieldNames;
        this.record = record;
    }

    @Override
    public T get(int index) {
        return getByIndex(index, record);
    }

    @Override
    public int getSize() {
        return record.getSize();
    }

    @Override
    public T get(String field) {
        if (fieldNames.containsKey(field)) {
            return getByIndex(fieldNames.get(field), record);
        }
        return null;
    }

    private T getByIndex(Integer index, Indexible<T> record) {
        if (index >= 0) {
            return record.get(index);
        } else {
            return record.get(record.getSize() + index);
        }
    }

    @Override
    public List<String> keys() {
        return fieldNames.keySet().stream().collect(Collectors.toList());
    }

    @Override
    public String toString() {
        return record.toString();
    }
}

File: rec-accessor/src/main/kotlin/net/kimleo/rec/accessor/RecordWrapper.kt
package net.kimleo.rec.accessor

import net.kimleo.rec.concept.Indexible
import net.kimleo.rec.concept.Mapped

data class RecordWrapper<out T>(val fieldNames: Map<String, Int>, val record: Indexible<T>): Indexible<T>, Mapped<T> {
    override fun keys(): List<String> {
        return fieldNames.keys.toList()
    }

    override operator fun get(field: String): T? {
        if (fieldNames.containsKey(field)) {
            return getByIndex(fieldNames[field]!!, record)
        }
        return null
    }

    override fun get(index: Int): T {
        return record[index]
    }

    override val size = record.size

    private fun getByIndex(index: Int, record: Indexible<T>): T? {
        if (index >= 0) {
            return record[index]
        } else {
            return record[record.size + index]
        }
    }
}